lib/ansible/plugins/test/success.yml

```diff
@@ -5,8 +5,9 @@ DOCUMENTATION:
   short_description: check task success
   aliases: [succeeded, successful]
   description:
-    - Tests if task finished successfully, opposite of C(failed).
+    - Test if task did not fail; returning the opposite of C(failed).
```

Comment:
I  think the original phrasing reads better and is equivalent in meaning.

Reply:
If you propose to say that the task "finished successfully", there would need to be heavy qualifications about what those words mean, explaining how it can be that a task that "finished successfully" did not start running on the host, therefore it could not have finished running on the host, and it could not have succeeded on the host, because it was skipped.

@s-hertel in the case of a skipped task, when nothing has been executed on the target host, can you explain exactly what thing/event/process finished successfully? Do we want to say e.g. that it "finished successfully being processed by the Ansible task engine", to draw a clear distinction against task execution on the target host?

Reply:
Success is the opposite of failed. A dictionary is considered 'failed' when it contains the key 'failed' and the value is True. A skipped task never contains the key 'failed', therefore a skipped task is successful, no qualifier needed.

Reply:
All skipped tasks are successful but not all successful tasks are skipped. You were right in the previous issue, that the correct pattern to check if a host executed the task is with the success + skipped tests. I think adding something like https://github.com/ansible/ansible/pull/83705#pullrequestreview-2213716186 or my suggestions below would help clarify, without requiring rewording this line.

Reply:
If you had the assumption action == task, then I can understand why it might not be intuitive that a skipped action is success (skipped, success, and failed could have no overlap at all, for example).

Reply:
@s-hertel That is a circular definition. I asked, in the case of a skipped task, what precise event or occurrence completes or is successful on a host, and you said effectively that it succeeds because it is `ansible.builtin.successful`. In other words, "`ansible.builtin.success` tests if task returns `ansible.builtin.success`".  A circular definition, of course, does not define anything.

Aside from a clear technical definition (i.e. "returns the opposite of failed"), the definition should provide an explanation in plain simple language that can be easily interpreted by end users using the established meaning of basic English words like [success](https://www.dictionary.com/browse/success) or [complete](https://www.dictionary.com/browse/complete). E.g. "the task ran on a host successfully or was skipped" or "processing of the task on the Ansible controller host successfully completed without an error".

So if you propose to say that `ansible.builtin.success` tests if a task finished successfully, can you please explain what precise event or occurrence finishes and is successful in the case of a skipped task, which does not start, run, or succeed on a host, using the standard English meaning of "finished" and "successful", without resorting to a circular definition?   Then I will be glad to write the documentation to clarify this based on your explanation.

I don't see how the distinction between a task and action is material. 

> therefore a skipped task is successful, no qualifier needed.

The issue is basically considering `ansible.builtin.success` as interchangeable with [success](https://www.dictionary.com/browse/success), when in practice the result of `ansible.builtin.success` is  counterintuitive and at odds with a common sense expectation of what "successfully completed" means in the case of a skipped task - if it successfully completed it should have executed the operation on the target host and ensured that the result was as expected (i.e. the green or orange colour coded output indicating successful completion, not teal or red indicating skipped or failed). The meanings of the English words "success" and "complete" are not really open to different interpretations, so to avoid ambiguity, contradictions, and misinterpretations, if those words are used in the documentation they should be carefully and precisely qualified to clarify what exactly has succeeded and completed and how.

Reply:
simple: the task did not end in failure

Reply:
@bcoca Great, sounds good to me, we shall let the documentation say "Test if task did not end in failure". 
That is accomplished with this change I have proposed here,  although I can change it from "did not fail" to "did not end in failure" if you prefer.

Reply:
@rptaylor Register a skipped task and debug the registered variable - you will see it is a dictionary which does not contain 'failed'. If it's not failed, it's successful. Not circular, it's literally how it's written in the code.

The way you've phrased is redundant. Just delete the first part of the sentence instead of saying the same thing twice and using an unnecessary semicolon. (Though, I prefer the original)

Reply:
@s-hertel Yes, the technical details are clear, but the purpose of documentation is to explain in plain clear language how the application works so that you don't need to read the code.

> If it's not failed, it's successful. Not circular, it's literally how it's written in the code.

The crux of this issue is semantics - a skipped task is [successful](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/success_test.html) but it is not [successful](https://www.dictionary.com/browse/successful). Simply writing a word in code does not necessarily imbue the code with the meaning of that word. When there is a disconnection between the two it is especially important to highlight in documentation.

I will follow up with a revision after I hear back in the other thread. Thanks!

Reply:
A skipped task is [successful](https://www.merriam-webster.com/dictionary/successful). A [task](https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/task.py#L46) is an ansible-core construct that interacts with a bunch of other ansible-core constructs. A skipped task is treated as successful during the ansible runtime - it doesn't trigger rescue blocks or max_fail_percentage, or show up as an error in your play recap or give a non-zero rc.